### PR TITLE
Defaults in TranspileConfig (now called PassManagerConfig)

### PR DESCRIPTION
**Pull request by 1ucian0**
While write the release notes for #3004, I realized editing a preset passmanager is a real use-case. That requires a more user-friendly `TranspileConfig`. For example, it needs good defaults.

If I understand correctly this use-case, a user should be able to instantiate a preset pass manager doing something like this:

```python
from qiskit.transpiler.preset_passmanagers.level0 import level_0_pass_manager
from qiskit.transpiler.transpile_config import TranspileConfig
pass_manager = level_0_pass_manager(TranspileConfig())
pass_manager.draw()
```

If the `level_0_pass_manager` parameter is not optional, then `TranspileConfig()` should be a thing. This PR allows it.

### CODE REVIEW

File: qiskit/transpiler/transpile_config.py [hunk 1]
```diff
@@ -30,6 +30,9 @@ class TranspileConfig(BaseModel):
             optimization level. 0 means no transformation on the circuit. Higher
             levels may produce more optimized circuits, but may take longer.
     """
-    def __init__(self, optimization_level, **kwargs):
+    def __init__(self, optimization_level=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (ajavadia) [reply qiskit/transpiler/transpile_config.py]: and let's expand kwargs to be args with default set to None. So that at least the structure of this class is clear and not just anything can go inside it.

**Comment by 1ucian0** [comment]
I split `transpiler_config` from `pass_manager_config`. This is the structure:
```python 
transpile_config = {PassManagerConfig(basis_gates,
                                      coupling_map,
                                      backend_properties,
                                      initial_layout,
                                      seed_transpiler,
                                      callback),
                    optimization_level,
                    pass_manager,
                    output_name,
                    circuit,
                    callback}
```

File: qiskit/compiler/transpile.py [hunk 2]
```diff
@@ -178,49 +178,21 @@ def callback_func(**kwargs):
...
-    circuits = parallel_map(_transpile_circuit, list(zip(circuits, transpile_configs)))
+    circuits = parallel_map(transpile_circuit, transpile_args)
 
     if len(circuits) == 1:
         return circuits[0]
     return circuits
 
 
-# FIXME: This is a helper function because of parallel tools.
-def _transpile_circuit(circuit_config_tuple):
```
Comment (kdk) [inline qiskit/compiler/transpile.py]: Did something change that this is no longer needed (that circuits are now in `transpile_args`), or was it always unnecessary?

File: qiskit/compiler/transpile.py [hunk 3]
```diff
@@ -234,42 +206,36 @@ def _parse_transpile_args(circuits, backend,
...
-                                           coupling_map=args[1],
-                                           backend_properties=args[2],
-                                           initial_layout=args[3],
-                                           seed_transpiler=args[4],
-                                           optimization_level=args[5],
-                                           pass_manager=args[6],
-                                           callback=callback,
-                                           output_name=args[7])
+                    pass_manager, output_name, circuits):
+        transpile_config = {'pass_manager_config': PassManagerConfig(basis_gates=args[0],
```
Comment (kdk) [inline qiskit/compiler/transpile.py]: `transpile_args`?

File: qiskit/compiler/transpile.py [hunk 4]
```diff
@@ -234,42 +206,36 @@ def _parse_transpile_args(circuits, backend,
...
+                                                                     coupling_map=args[1],
+                                                                     backend_properties=args[2],
+                                                                     initial_layout=args[3],
+                                                                     seed_transpiler=args[4],
+                                                                     callback=callback),
+                            'optimization_level': args[5],
+                            'pass_manager': args[6],
+                            'output_name': args[7],
+                            'circuit': args[8],
+                            'callback': callback}
```
Comment (kdk) [inline qiskit/compiler/transpile.py]: Is `callback` needed in both `transpile_args` and `PassManagerConfig`?

File: qiskit/compiler/transpile.py [hunk 5]
```diff
@@ -301,18 +267,29 @@ def _parse_basis_gates(basis_gates, backend, circuits):
...
-        coupling_map = [coupling_map] * num_circuits
-    elif isinstance(coupling_map, list) and all(isinstance(i, list) and len(i) == 2
-                                                for i in coupling_map):
-        coupling_map = [coupling_map] * num_circuits
-    coupling_map = [CouplingMap(cm) if isinstance(cm, list) else cm for cm in coupling_map]
+
+    coupling_map = [coupling_map] * num_circuits
+
+    for index, parsed_coupling_map in enumerate(coupling_map):
+        if isinstance(parsed_coupling_map, list):
```
Comment (kdk) [inline qiskit/compiler/transpile.py]: Could we maybe save a little work by doing this once from the input `coupling_map` before multiplying by `num_circuits`?

File: qiskit/transpiler/passmanager.py [hunk 6]
```diff
@@ -152,11 +152,16 @@ def reset(self):
         self.valid_passes = set()
         self.property_set.clear()
 
-    def run(self, circuit):
+    def run(self, circuit, output_name=None, callback=None):
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/passmanager.py]: Yes, if for reusability. Given a particular `PassManager` you might want to add a callback for debugging.

File: qiskit/transpiler/pass_manager_config.py [hunk 7]
```diff
@@ -12,24 +12,35 @@
...
+@bind_schema(PassManagerConfigSchema)
+class PassManagerConfig(BaseModel):
+    """Model for PassManagerConfig.
 
     Please note that this class only describes the required fields. For the
-    full description of the model, please check ``TranspileConfigSchema``.
+    full description of the model, please check ``PassManagerConfigSchema``.
 
     Attributes:
         optimization_level (int): a non-negative integer indicating the
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/pass_manager_config.py]: `transpile_args` is a dict because is only used for calling the parallelization code.

File: qiskit/transpiler/passmanager.py [hunk 8]
```diff
@@ -166,13 +171,21 @@ def run(self, circuit):
...
         for passset in self.working_list:
             for pass_ in passset:
                 dag = self._do_pass(pass_, dag, passset.options)
 
         circuit = dag_to_circuit(dag)
-        circuit.name = name
+
+        if output_name:
+            circuit.name = output_name
+        else:
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/passmanager.py]: you might call `PassManager.run` without calling transpiler. Maybe is better for `_parse_output_name` to set `None` when no change?

File: qiskit/compiler/transpile.py (hunk 2 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: My understanding of `FIXME` is that it was not needed.

File: qiskit/compiler/transpile.py (hunk 3 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: I will refactor all the `transpile_config` to `transpile_args` for consistency. Thanks.

File: qiskit/compiler/transpile.py (hunk 5 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: They will all have the same instance of coupling map... I'm thinking over the consequence of that. But it might be something positive...

File: CHANGELOG.md [hunk 9]
```diff
@@ -17,6 +17,10 @@ The format is based on [Keep a Changelog].
 
 ## [0.9.0] - 2019-08-22
 
+### Changed
+
+-   The `TranspileConfig` was extended to have `None` defaults for
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply CHANGELOG.md]: fixed in 01de59bbc8291e1a3243d1b45fafebc4a7eb6b4c

**Comment by 1ucian0** [comment]
After fe54db62572bd095ef85c52e47d48794c8cd235d :
```
transpile_args = {PassManagerConfig(basis_gates,
                                      coupling_map,
                                      backend_properties,
                                      initial_layout,
                                      seed_transpiler),
                  optimization_level,
                  pass_manager,
                  output_name,
                  circuit,
                  callback}
```

File: qiskit/transpiler/passmanager.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/passmanager.py]: After some discussion with @mtreinish and @kdk we are going to keep callback at run-time and deprecated its use at construction-time.

File: qiskit/transpiler/passmanager.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/passmanager.py]: Fixed in 26b379e6

File: qiskit/compiler/transpile.py (hunk 3 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: Fixed in 4aeb87ef

File: qiskit/compiler/transpile.py (hunk 4 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: removed in fe54db62572bd095ef85c52e47d48794c8cd235d

File: qiskit/compiler/transpile.py (hunk 5 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: this situation handles when the input `coupling_map` is a list of coupling maps defined as a list. Therefore, they might be different among them. In 948b3d44, I construct a coupling map once if it's coming from the backend `CouplingMap(backend.configuration().coupling_map)`

**Comment by 1ucian0** [comment]
Closed in favor of https://github.com/Qiskit/qiskit-terra/pull/3290/

**Comment by kdk** [comment]
Can you add a release note for the info that had been in the changelog?

**Comment by 1ucian0** [comment]
> Can you add a release note for the info that had been in the changelog?

Done in https://github.com/Qiskit/qiskit-terra/pull/3035/commits/6b0b4bec6bb611eec47bcfb1352754016f07ee1b

File: qiskit/transpiler/transpile_circuit.py [hunk 10]
```diff
@@ -22,61 +22,57 @@
...
 
 
-def transpile_circuit(circuit, transpile_config):
+def transpile_circuit(transpile_args):
     """Select a PassManager and run a single circuit through it.
 
     Args:
-        circuit (QuantumCircuit): circuit to transpile
-        transpile_config (TranspileConfig): configuration dictating how to transpile
+        transpile_args (dict): configuration dictating how to transpile
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/transpile_circuit.py]: The way that parallel works is to call each `apply_async` with task and each element of the `value` dict. So `transpile_circuit` literally receives a dict as a single argument. `transpile_args` is already `zip`'ed by `_parse_transpile_args`. `parallel_map` needs another thought maybe. The change needs a proper refactor for a following PR.

I updated the docstring in ad5bb3a7e

File: releasenotes/notes/defaults_TranspileConfig-4109a90c278d46df.yaml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+---
+other:
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply releasenotes/notes/defaults_TranspileConfig-4109a90c278d46df.yaml]: Fixed in b21e0b35f

File: releasenotes/notes/defaults_TranspileConfig-4109a90c278d46df.yaml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+---
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply releasenotes/notes/defaults_TranspileConfig-4109a90c278d46df.yaml]: Fixed in 70aaf6c21

File: qiskit/transpiler/pass_manager_config.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ajavadia) [reply qiskit/transpiler/pass_manager_config.py]: attributes documentation should be updated.

File: releasenotes/notes/defaults_TranspileConfig-4109a90c278d46df.yaml [hunk 13]
```diff
@@ -0,0 +1,9 @@
...
+upgrade:
+  - |
+    The ``TranspileConfig`` was restructured to include only the information
+    needed to construct a ``PassManager`` and it was renamed to ``PassManagerConfig``.
```
Comment (ajavadia) [inline releasenotes/notes/defaults_TranspileConfig-4109a90c278d46df.yaml]: > only the information needed to construct a ``PassManager`` 

can you elaborate on this

File: qiskit/transpiler/transpile_circuit.py [hunk 14]
```diff
@@ -22,61 +22,64 @@
 from qiskit.transpiler.exceptions import TranspilerError
 
 
-def transpile_circuit(circuit, transpile_config):
+def transpile_circuit(transpile_args):
```
Comment (ajavadia) [inline qiskit/transpiler/transpile_circuit.py]: passing in a dict here (`transpile_args`) does not look like a good API. 

try expanding it:
`transpile_circuit(circuit, pass_manager, pass_manager_config, output_name, callback)`
and give them defaults

EDIT: i just saw @mtreinish made a similar comment.

File: qiskit/transpiler/runningpassmanager.py [hunk 15]
```diff
@@ -120,25 +92,33 @@ def _normalize_flow_controller(self, flow_controller):
...
+            callback (callable): A callback function that will be called after each pass execution.
         Returns:
             QuantumCircuit: Transformed circuit.
         """
         name = circuit.name
         dag = circuit_to_dag(circuit)
         del circuit
 
+        if callback:
+            self.callback = callback
```
Comment (ajavadia) [inline qiskit/transpiler/runningpassmanager.py]: i don't get how callback works in this RunningPassManager class. you just set `self.callback`, and nothing else seems to happen.

File: qiskit/transpiler/passmanager.py [hunk 16]
```diff
@@ -207,29 +214,61 @@ def _in_parallel(circuit, pm_dill=None):
...
-
-        Args:
-            circuits (list[QuantumCircuit]): circuit to transform via all the registered passes
-
         Returns:
             list[QuantumCircuit]: Transformed circuits.
         """
+        # TODO support for List(output_name) and List(callback)
+        del output_name
+        del callback
```
Comment (ajavadia) [inline qiskit/transpiler/passmanager.py]: why are these deleted?

File: qiskit/transpiler/transpile_circuit.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ajavadia) [reply qiskit/transpiler/transpile_circuit.py]: the way `parallel_map` works is an implementation detail and it should not end up dictating the user-facing API. if you need to pass it a dict, then just internally make a dict for that sake only.

**Comment by ajavadia** [comment]
Since it took me a while to see what's going on, I'll summarize my findings. The release notes should have more details.

The original `TranspileConfig` contained the following fields:
```
initial_layout,
basis_gates,
coupling_map,
backend_properties,
seed_transpiler,

optimization_level,
pass_manager,
callback,
output_name
```

the new `PassManagerConfig` has just:
```
initial_layout=None,
basis_gates=None,
coupling_map=None,
backend_properties=None,
seed_transpiler=None,
```
now with None defaults. This makes it easier to instantiate a preset passmanager:

```
from qiskit.transpiler.preset_passmanagers.level0 import level_0_pass_manager
from qiskit.transpiler.passmanager_config import PassManagerConfig
pass_manager = level_0_pass_manager(TranspileConfig())
```

The other attributes are moved to a "transpile_args" dictionary, which I don't really like. They should all just be expanded in the functions that accept them. See comments above.

File: qiskit/transpiler/transpile_circuit.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/transpile_circuit.py]: Is `transpile_circuit` user-facing API?

[the remark this answers is not in the document]
Comment (ajavadia) [reply qiskit/transpiler/transpile_circuit.py]: well it's not a private function and it's in `qiskit.transpiler.__init__`. But the point is I don't think this change is necessary just to accommodate the needs of `parallel_map`. That can be handled internally.

[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/transpile_circuit.py]: Can we remove it ? How is it different from `transpile`? I can reintroduce `_transpile_circuit`.

**Comment by 1ucian0** [comment]
```python
from qiskit.transpiler.preset_passmanagers.level0 import level_0_pass_manager
from qiskit.transpiler.passmanager_config import PassManagerConfig
pass_manager = level_0_pass_manager(TranspileConfig())
```
should be
```python
from qiskit.transpiler.preset_passmanagers.level0 import level_0_pass_manager
from qiskit.transpiler.passmanager_config import PassManagerConfig
pass_manager = level_0_pass_manager(PassManagerConfig())
```

File: qiskit/transpiler/transpile_circuit.py (hunk 14 above)
Comment (1ucian0) [reply qiskit/transpiler/transpile_circuit.py]: I'm returning back to the previous situation. Making that change would be changing the API...

File: qiskit/transpiler/passmanager.py (hunk 16 above)
Comment (1ucian0) [reply qiskit/transpiler/passmanager.py]: I will add support for List(output_name) and List(callback) in a coming PR. I will remove these del after that.

File: qiskit/transpiler/transpile_circuit.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/transpile_circuit.py]: If `transpile_circuit` is API, I'm rolling back to how it was. I would like to deprecate it in a coming PR.

File: qiskit/transpiler/runningpassmanager.py (hunk 15 above)
Comment (1ucian0) [reply qiskit/transpiler/runningpassmanager.py]: This makes possible to set a callback at passmanager running-time.

File: qiskit/compiler/transpile.py (hunk 2 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: I introduced it back.

File: releasenotes/notes/defaults_TranspileConfig-4109a90c278d46df.yaml (hunk 13 above)
Comment (1ucian0) [reply releasenotes/notes/defaults_TranspileConfig-4109a90c278d46df.yaml]: I added the fields 34aec1e29. You think something else should be added?

File: qiskit/transpiler/transpile_circuit.py (hunk 14 above)
Comment (ajavadia) [reply qiskit/transpiler/transpile_circuit.py]: This is not the old API, it's now a dict rather than a TranspileConfig.

But that's not what I care about. My comment wasn't about "let's not change API". My point was "let's not have a hard-to-use API". I think a dict is hard to use. If we know the keys of the dict, might as well expand it into kwargs.

File: qiskit/transpiler/pass_manager_config.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ajavadia) [reply qiskit/transpiler/pass_manager_config.py]: Can you make sure the module docstring has all the right info about what this new class contains.

File: qiskit/transpiler/transpile_circuit.py (hunk 14 above)
Comment (1ucian0) [reply qiskit/transpiler/transpile_circuit.py]: This snippet is outdate. Check here https://github.com/Qiskit/qiskit-terra/pull/3035/files#diff-751f19949d593b00af720b607ee5f7e5R213 

I want to put it back, because is an irrelevant change. I'm thinking on deprecating this function in a coming PR.

Comment (ajavadia) [reply qiskit/transpiler/transpile_circuit.py]: I'm referring to the current up to date version. You have changed TranspileConfig to a dict, so it is an API change. And I'm referring to this dict when I say this is a bad API.
```python
        transpile_config (dict): configuration dictating how to transpile.
            The format of the dictionary has the following format:
            {'circuit': QuantumCircuit,
             'optimization_level': int,
             'pass_manager': PassManager,
             'output_name': string,
             'callback': callable,
             'pass_manager_config': PassManagerConfig}
```

Comment (1ucian0) [reply qiskit/transpiler/transpile_circuit.py]: I see. Making that change. Can I also deprecate `transpile_circuit`? Or that is for a different PR?

**Comment by ajavadia** [comment]
Can you try and see if, using this PR, we can make progress on solving #3118

**Comment by 1ucian0** [comment]
> Can you try and see if, using this PR, we can make progress on solving #3118

I see this PR in the direction of #3118, but solving #3118 will create a lot of changes. I can add it in my backlog for doing next.

File: qiskit/compiler/transpile.py [hunk 17]
```diff
@@ -201,28 +207,72 @@ def callback_func(**kwargs):
...
-# FIXME: This is a helper function because of parallel tools.
 def _transpile_circuit(circuit_config_tuple):
     """Select a PassManager and run a single circuit through it.
-
     Args:
         circuit_config_tuple (tuple):
             circuit (QuantumCircuit): circuit to transpile
-            transpile_config (TranspileConfig): configuration dictating how to transpile
-
+            transpile_config (dict): configuration dictating how to transpile. The format of the
```
Comment (ajavadia) [inline qiskit/compiler/transpile.py]:
```suggestion
            transpile_config (dict): configuration dictating how to transpile. The
```

File: qiskit/compiler/transpile.py [hunk 18]
```diff
@@ -201,28 +207,72 @@ def callback_func(**kwargs):
...
+                dictionary has the following format:
+                {'optimization_level': int,
+                 'pass_manager': PassManager,
+                 'output_name': string,
+                 'callback': callable,
+                 'pass_manager_config': PassManagerConfig}
```
Comment (ajavadia) [inline qiskit/compiler/transpile.py]: make sure all docstrings are accurate. this one still has optimization_level and pass_manager_config in it.

File: qiskit/transpiler/transpile_circuit.py (hunk 14 above)
Comment (ajavadia) [reply qiskit/transpiler/transpile_circuit.py]: Yeah I don't think we need to deprecate that.

File: qiskit/transpiler/passmanager.py (hunk 16 above)
Comment (ajavadia) [reply qiskit/transpiler/passmanager.py]: can you leave a TODO comment then

File: qiskit/transpiler/transpile_circuit.py (hunk 14 above)
Comment (1ucian0) [reply qiskit/transpiler/transpile_circuit.py]: Can I remove `transpile_circuit` directly? Or are you talking about the parameters?

File: qiskit/transpiler/passmanager.py (hunk 16 above)
Comment (1ucian0) [reply qiskit/transpiler/passmanager.py]: It's in line 222.

File: qiskit/compiler/transpile.py (hunk 18 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: Because this is `_transpile_circuit`, not `transpile_circuit`. Changing it here implies modifying `_parse_transpile_args`. It is necessary, but this PR is already hard to understand...

File: qiskit/transpiler/pass_manager_config.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ajavadia) [reply qiskit/transpiler/pass_manager_config.py]: this is still not addressed. docstrings are not up to date

File: qiskit/transpiler/transpile_circuit.py [hunk 19]
```diff
@@ -14,69 +14,22 @@
...
+                      output_name: None,
+                      callback: None):
     """Select a PassManager and run a single circuit through it.
-
     Args:
         circuit (QuantumCircuit): circuit to transpile
-        transpile_config (TranspileConfig): configuration dictating how to transpile
+        pass_manager (PassManager): The pass manager to use for a custom pipeline of
+            transpiler passes.
+        output_name (string): To identify the output circuits
```
Comment (ajavadia) [inline qiskit/transpiler/transpile_circuit.py]:
```suggestion
        output_name (string): name of output circuit
```

File: qiskit/compiler/transpile.py [hunk 20]
```diff
@@ -238,45 +288,38 @@ def _parse_transpile_args(circuits, backend,
...
-
     optimization_level = _parse_optimization_level(optimization_level, num_circuits)
-
     pass_manager = _parse_pass_manager(pass_manager, num_circuits)
+    output_name = _parse_output_name(output_name, num_circuits)
 
-    output_name = _parse_output_name(output_name, circuits)
-
-    transpile_configs = []
+    transpile_args_circuits = []
```
Comment (ajavadia) [inline qiskit/compiler/transpile.py]: suggest better name, as there are no circuits involved in the dict
```suggestion
    list_transpile_args = []
```

File: qiskit/compiler/transpile.py [hunk 21]
```diff
@@ -238,45 +288,38 @@ def _parse_transpile_args(circuits, backend,
...
+                    pass_manager, output_name, circuits):
+        transpile_args = {'pass_manager_config': PassManagerConfig(basis_gates=args[0],
+                                                                   coupling_map=args[1],
+                                                                   backend_properties=args[2],
+                                                                   initial_layout=args[3],
+                                                                   seed_transpiler=args[4]),
+                          'optimization_level': args[5],
+                          'pass_manager': args[6],
+                          'output_name': args[7],
+                          'circuit': args[8],
```
Comment (ajavadia) [inline qiskit/compiler/transpile.py]: the circuit should not be part of the transpile_args (see line 210)

File: qiskit/compiler/transpile.py [hunk 22]
```diff
@@ -238,45 +288,38 @@ def _parse_transpile_args(circuits, backend,
...
+                          'callback': callback}
```
Comment (ajavadia) [inline qiskit/compiler/transpile.py]: callback should be like other arguments here.

File: qiskit/transpiler/pass_manager_config.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/transpiler/pass_manager_config.py]: Fixed in fa007a0da

File: qiskit/compiler/transpile.py (hunk 22 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: Fixed in 6ed967776

File: qiskit/compiler/transpile.py (hunk 21 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: Good catch! Fixed in 7ec5cebc2

File: qiskit/compiler/transpile.py (hunk 20 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: Sure... fixed in 4c87790db

File: qiskit/transpiler/transpile_circuit.py [hunk 23]
```diff
@@ -14,69 +14,22 @@
...
-from qiskit.transpiler.passes.basis.ms_basis_decomposer import MSBasisDecomposer
-from qiskit.transpiler.exceptions import TranspilerError
 
-
-def transpile_circuit(circuit, transpile_config):
+def transpile_circuit(circuit,
+                      pass_manager,
+                      output_name: None,
+                      callback: None):
     """Select a PassManager and run a single circuit through it.
```
Comment (ajavadia) [inline qiskit/transpiler/transpile_circuit.py]: This no longer "selects" a PassManager. The PassManager is given. 

Again, please thoroughly read all the docstrings for the affected modules and functions because a lot of old behavior is changing here.

Comment (ajavadia) [reply qiskit/transpiler/transpile_circuit.py]: I actually think this whole file can be removed. The point of this `transpile_circuit()` function was to provide a thin layer of abstraction over `PassManager.run()`, and handle the selection of pass manager. Now the pass manager is given, and this function does nothing but call `PassManager.run()`. 

You just have to make parallel_map call `PassManager.run(circuit)` rather than `transpile_circuit(circuit)`

File: qiskit/compiler/transpile.py [hunk 24]
```diff
@@ -299,18 +342,31 @@ def _parse_basis_gates(basis_gates, backend, circuits):
     return basis_gates
 
 
-def _parse_coupling_map(coupling_map, backend, num_circuits):
+def _parse_coupling_map(coupling_map, backend, circuits):
```
Comment (ajavadia) [inline qiskit/compiler/transpile.py]: why is this function changing at all in this PR? It's already a bit hard for me to follow what's going on, and this one seems completely unrelated. 

Same for `_parse_output_name()`

File: qiskit/compiler/transpile.py [hunk 25]
```diff
@@ -201,28 +207,72 @@ def callback_func(**kwargs):
...
+    basic_insts = ['measure', 'reset', 'barrier', 'snapshot']
+    device_insts = set(pass_manager_config.basis_gates).union(basic_insts)
+    ms_basis_swap = None
+    if 'rxx' in pass_manager_config.basis_gates and \
+            not device_insts >= circuit.count_ops().keys():
+        ms_basis_swap = pass_manager_config.basis_gates
+        pass_manager_config.basis_gates = list(
+            set(['u3', 'cx']).union(pass_manager_config.basis_gates))
+
+    if transpile_config['pass_manager'] is not None:
```
Comment (ajavadia) [inline qiskit/compiler/transpile.py]: there were some code comments here that got lost during copy/paste. comments are good :)

File: qiskit/compiler/transpile.py (hunk 18 above)
Comment (ajavadia) [reply qiskit/compiler/transpile.py]: ok, can you rename `transpile_config` -> `transpile_args` then to avoid confusion with the deprecated class of the same name?

File: qiskit/compiler/transpile.py (hunk 24 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: Sure.. the goal was to encapsulate the cm checks here. But I can do that in a following PR. Undone in e1d7f10da for `_parse_coupling_map`.

For the `_parse_output_name` case, there was unnecessary logic there, because `None` means "keep the input name". Undone in 7ad1521f2.

File: qiskit/transpiler/transpile_circuit.py (hunk 23 above)
Comment (1ucian0) [reply qiskit/transpiler/transpile_circuit.py]: `transpile_circuit` is part of the API, I can remove in a coming PR. The docstring was updated here: 346b8ea21

File: qiskit/compiler/transpile.py (hunk 25 above)
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: commnets in b84e270ac83a1dbfc3d493da7abe69751843f8bd

File: qiskit/transpiler/transpile_circuit.py (hunk 23 above)
Comment (ajavadia) [reply qiskit/transpiler/transpile_circuit.py]: `transpile_circuit()` has already changed in this PR. It used to take a `TranspileConfig` and no longer does.

Comment (1ucian0) [reply qiskit/transpiler/transpile_circuit.py]: Done in 8d935be56

**Comment by mtreinish** [comment]
I agree with @ajavadia lets make this an early merge for the 0.13.0 cycle. I'm updating the milestone accordingly.

**Comment by 1ucian0** [comment]
Removing `on hold` since there was a release.
